hpx/parallel/algorithms/is_heap.hpp

```diff
@@ -0,0 +1,201 @@
+#include <iterator>
+#include <list>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
+{
+    ///////////////////////////////////////////////////////////////////////////
+    // is_heap_until
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+        struct is_heap_until_helper
+        {
+            template <typename ExPolicy, typename RandIter, typename Comp>
+            typename util::detail::algorithm_result<ExPolicy, RandIter>::type
+            operator()(ExPolicy && policy, RandIter first, RandIter last, Comp comp)
+            {
+                typedef util::detail::algorithm_result<ExPolicy, RandIter> result;
+                typedef typename std::iterator_traits<RandIter>::value_type type;
+                typedef typename std::iterator_traits<RandIter>::difference_type
+                    difference_type;
+
+                difference_type count = std::distance(first, last);
+                if (count <= 1)
+                    return result::get(std::move(last));
+
+                RandIter second = first + 1;
+                --count;
+                util::cancellation_token<std::size_t> tok(count);
+
+                return util::partitioner<ExPolicy, RandIter, void>::
+                    call_with_index(
+                        std::forward<ExPolicy>(policy), second, count, 1,
+                        [tok, first, comp](RandIter it, std::size_t part_size,
+                            std::size_t base_idx) mutable
+                        {
+                            util::loop_idx_n(
+                                base_idx, it, part_size, tok,
+                                [&tok, first, &comp](type const& v, std::size_t i)
+                                {
+                                    if (comp(*(first + i / 2), v))
+                                        tok.cancel(i);
+                                });
+                        },
+                        [tok, second](std::vector<hpx::future<void> > &&) mutable -> RandIter
+                        {
+                            difference_type find_res =
+                                static_cast<difference_type>(tok.get_data());
+                            
+                            std::advance(second, find_res);
+
+                            return std::move(second);
+                        });
+            }
+        };
+
+        template <typename RandIter>
+        struct is_heap_until
+          : public detail::algorithm<is_heap_until<RandIter>, RandIter>
+        {
+            is_heap_until()
+              : is_heap_until::algorithm("is_heap_until")
+            {}
+
+            template <typename ExPolicy, typename Comp>
+            static RandIter
+            sequential(ExPolicy && policy, RandIter first, RandIter last,
+                Comp && comp)
+            {
+                return std::is_heap_until(first, last, std::forward<Comp>(comp));
```

Comment:
Yah, we do it inconsistently throughout the algorithm implementations. Most of the time iterators are passed by value, sometimes not. The Standard prescribes for the algorithms to take iterators by value (on the outermost layer), everything else is up to the implementation I guess.